### PR TITLE
DATUM ANTAGS

### PR DESCRIPTION
**Pull request by MacHac**
Many months ago, I had a vision.  I saw multiple modes running at once, each with their own objective tracking.  I saw order, where there is chaos.  I saw a system for tracking antagonists that actually made sense.

I am now determined to make this vision a reality.

The contents of this PR, at the time of this writing, are a highly incomplete and half baked system that will migrate us over from our current one-mode system and its piles and piles of global variables to a proper system that handles as many different antagonist types at once as we have and more, all in a clean and ordered format with things like _conventions_ and _consistency_.

Obviously, this is a task of astounding magnitude, since to my knowledge people have just been shoveling code on top of our broken-ass system for literally years without moving to fix it.  Once again, at the time of this writing, I've touched dozens of files and thousands of lines and I'm _maybe_ halfway done.  So, I'm putting this half-baked PR on the list and anyone who wants to crawl through it and poke holes in my half-baked 3AM programming can do so.

### MISSION STATEMENT
Ultimately, the goal is to replace most of the role of the game_mode datum with a new datum called game that can track multiple game_modes.  The game datum effectively standardizes the format of the game_modes by moving all of the mode's functions and variables to the object definition instead of just having 300 different variables in every game_mode regardless of need.

To compensate for this, it becomes possible to add and remove modes from a running game on-the-fly and without having to make special snowflake handler classes for it (here's looking at you, traitorductorling).  Previously, if an admin wanted to make cult v. cult with full objectives and such, it would be a huge issue and inevitably involve missing objectives, missing functionality, and hacky snowflake code.  Now, it's two button clicks.  Wanna add xenos into the mix?  Wanna make the AI malf?  Wanna drop a dozen blobs on the station and pit them against a nuke squad with exosuits and flamethrowers?  It just _works_.  This extends to the config file too.  Instead of just stating probabilities for one mode at a time, you can now do combinations of multiple modes for stress-free tweaking.

The big _new_ feature here (that isn't just unfucking what we should have unfucked years ago) is mode arguments.  These allow you to make small tweaks to the modes themselves without changing anything globally.  For example, if you want to make 1% of rounds spawn 3 wizards instead of 1, you don't need to make a new game mode OR add snowflake code.  Just drop it in the config file and you're set.  If an admin wants to force 3 wizards in, they have that option.  You can see examples of the implementation of this in the wizard and traitor classes, among others.  For traitor mode, the entire Double Agent 'mode' was combined with standard traitor.  Now, instead of 
PROBABILITY TRAITOR 10
PROBABILITY DOUBLE_AGENTS 7

You have
!modes
traitor-med:10
traitor-hi+da:7
modes!
Want less traitors in double agent?
traitor-med+da:7
Want changelings to spawn in double agent?
traitor-med+da;ling-low:7
Want to add wizards with double-agent style objectives to kill each other?
traitor-med+da;ling-low;wizard-duo:7
Want to add a third wizard to make a triangle of assassination objectives?
traitor-med+da;ling-low;wizard-trifecta:7

Bottom line is, this has been a long time coming.  Trouble is, the current game_mode setup is one of the more fundamental things in this entire game, and a truly absurd amount of things depend on it.  So, anyone who finds time in their busy schedules to scroll through the ravings of a madman looking for obvious and non-obvious bugs is very much encouraged to do so.
:cl:
tweak: Significantly changed the internal behavior of gamemodes.
/:cl:

### CODE REVIEW

**Comment by ShadowDeath6** [comment]
nice

**Comment by Daman453** [comment]
I want this to go though so bad, admin guided rounds are going to be so fun to play

**Comment by Cruix** [comment]
Everything in datum/game should probably just be in the ticker

**Comment by MacHac** [comment]
@Cruix It's an abstraction.  Personally, I prefer 2 600 line classes to a single 1200 line class, and on top of that having the second class lets all the gamemodes code stay in the same folder.

**Comment by Cruix** [comment]
You can keep them in different files but put them in the same class. If that is the only reason you want to have it, you may as well save a variable.

**Comment by Kmc2000** [comment]
Art.

**Comment by Dagdammit** [comment]
Godspeed, you beautiful bastard. (I lack anywhere near the code knowledge to be useful, sorry.)

**Comment by Frostben** [comment]
it sounds so fucking amazing it makes me tremble with excitement.

**Comment by matskuman5** [comment]
Misclick, apologies

**Comment by MacHac** [comment]
Update: Bug patching and updating is going slowly because byond only shows 100 exceptions at a time (and I've already dealt with at least twice that number).  I think I have a file finished, and a few dozen exceptions later I see more issues with it.

@Cruix It just feels like a nice QOL thing for the developers, for the low low price of one variable handle.  Plus, my way, if something screws the state of the game datum, then the entire ticker doesn't go down.

**Comment by ghost** [comment]
THE MADMAN

HE'S DONE IT

**Comment by Mrcreepssss** [comment]
@AdamElTablawy well he is call MadHack (almost) so i guess the name says it all. ;)

**Comment by ArgyleSweatshop** [comment]
I believe in you

**Comment by Super3222** [comment]
it'd probably be hell to transfer over xeno mode onto this, but then again it'll probably make it easier to create an alien genocide mode. two queens, 1 station.

**Comment by Super3222** [comment]
:cry:

**Comment by ShadowDeath6** [comment]
mac is dead, this pr killed him

**Comment by Super3222** [comment]
F

**Comment by Time-Green** [comment]
Are you still alive Mac?

**Comment by ShadowDeath6** [comment]
creeps scared him off
